cloud/migrate.js

```diff
@@ -0,0 +1,319 @@
+    }
+
+    if (def(handlers.afterSave) || def(handlers.migrateObject)) {
+      self._parse.Cloud.afterSave(klass, self.getAfterSave(klass));
+    }
+
+    if (def(handlers.beforeDelete) || def(handlers.migrateDelete)) {
+      self._parse.Cloud.beforeDelete(klass, self.getBeforeDelete(klass));
+    }
+
+    if (def(handlers.afterDelete)) {
+      self._parse.Cloud.afterDelete(klass, self.getAfterDelete(klass));
+    }
+  });
+
+  self._parse.Cloud.job("import", self.getImportJob());
+};
+
+Migrator.prototype.getBeforeSave = function(klass) {
+  var handlers = this.handlers(klass),
+    beforeSave = handlers.beforeSave,
+    migrate = handlers.migrateObject,
+    _Promise = this._parse.Promise;
+
+  return function(request, response) {
+    var obj;
+    // Skip beforeSave if it doesn't exist (duh) but also if this is just
+    // a quick second pass to keep people from worrying about not having
+    // an objectId in their migration.
+    var changed = request.object.dirtyKeys();
+    var shouldBeforeSave = def(beforeSave) &&
+      !(changed.length === 1 && changed[0] == consts.MIGRATION_KEY);
+
+    return _Promise.as().then(function() {
+      if (shouldBeforeSave) {
+        return beforeSave(request, ThisIsNotAResponseObject);
+      }
+    }).then(function(maybeNew) {
+      obj = isParseObject(maybeNew) ? maybeNew : request.object;
+
+      // Hybrid apps can explicitly opt-out.
+      if (obj.get(consts.MIGRATION_KEY) == consts.IS_MIGRATED || !migrate) {
+        return obj;
+      }
+
+      if (obj.isNew()) {
+        // Will re-dirty the object in afterSave so another beforeSave
+        // catches this case with the objectId available.
+        return obj;
+      }
+
+       // won't save unless the migation succeeds.
+      if (obj.get(consts.MIGRATION_KEY) == consts.NEEDS_SECOND_PASS) {
+        obj.set(consts.MIGRATION_KEY, consts.FINISHED_SECOND_PASS);
+      } else {
+        obj.set(consts.MIGRATION_KEY, consts.IS_MIGRATED);
+      }
+      return _Promise.as().then(function() {
+        return migrate && migrate(obj);
+      }).then(function() {
+        // migration functions shouldn't change the object or it will
+        // behave weirdly between beforeSave and the migration job.
+        return obj;
+      });
+
+    }).then(function() {
+      return response.success(obj);
+    }, function(err) {
+      return response.error(err);
+    });
+  }
+};
+
+Migrator.prototype.getAfterSave = function(klass) {
+  var handlers = this.handlers(klass),
+    afterSave = handlers.afterSave,
+    migrate = handlers.migrateObject,
+    _Promise = this._parse.Promise,
+    maybeTouch = function(obj) {
+      if (obj.existed() || !def(migrate)) {
+        return _Promise.as();
+      }
+      obj.set(consts.MIGRATION_KEY, consts.NEEDS_SECOND_PASS);
+      return obj.save();
+    }
+
+  return function(request) {
+    // We don't have long and there is no failure mode. Let's do this in parallel:
+    return def(afterSave) ?
```

Comment:
afterSave never changes once this function is returned, so I think it would be clearer to write this part as:

    if (_.isUndefined(afterSave)) {
      return function(req) { maybeTouch(request.object); };
    }
    return function(request) {
      // We don't have long and there is no failure mode. Let's do this in parallel:
      _Promise.all([afterSave(request), maybeTouch(request.object)]);
    };